src/libraries/filcrypto/filproofs/algorithms.go

```diff
@@ -971,42 +1024,37 @@ func computePartialTicket(randomness sector.PoStRandomness, sectorID sector.Sect
 	return partialTicket
 }
 
-func (sdr *WinStackedDRG_I) _generatePoStCandidates(challengeSeed sector.PoStRandomness, eligibleSectors []sector.SectorID, candidateCount int, sectorStore sector_index.SectorStore) (candidates []sector.PoStCandidate) {
-
-	nodes := int(sdr.Nodes())
-	leafChallengeCount := int(sdr.LeafChallengeCount())
-	challengeRangeSize := int(sdr.ChallengeRangeSize())
-	challengedSectors := sdr._getChallengedSectors(eligibleSectors, challengeSeed, eligibleSectors, candidateCount)
-
-	for i, sectorID := range challengedSectors {
-		proofAux := sectorStore.GetSectorPersistentProofAux(sectorID)
+type PoStCandidatesMap map[sector.SealAlgorithm][]sector.PoStCandidate
 
-		candidate := generateCandidate(challengeSeed, proofAux, sectorID, UInt(i), leafChallengeCount, nodes, challengeRangeSize)
+func CreatePoStProof(cfg sector.PoStCfg, privateCandidateProofs []PrivatePostCandidateProof, challengeSeed sector.PoStRandomness) []sector.PoStProof {
+	var proofsMap map[sector.SealAlgorithm][]PrivatePostCandidateProof
 
-		candidates = append(candidates, candidate)
+	for _, proof := range privateCandidateProofs {
+		algorithm := proof.Algorithm()
+		proofsMap[algorithm] = append(proofsMap[algorithm], proof)
 	}
 
-	return candidates
-}
-
-func (sdr *WinStackedDRG_I) CreatePoStProof(privateCandidateProofs []sector.PrivatePoStCandidateProof, challengeSeed sector.PoStRandomness) sector.PoStProof {
-	privateProof := sdr.CreatePrivatePoStProof(privateCandidateProofs, challengeSeed)
+	var circuitProofs []sector.PoStProof
+	for algorithm, proofs := range proofsMap {
+		privateProof := createPrivatePoStProof(algorithm, proofs, challengeSeed)
+		// Hmmmmm: we cannot perform this santiy check without the sectorIDs. Should we require them just for that purpose or perform the check earlier?
```

Comment:
The proposed verification is of the private proof which is an intermediate product in creation of the `PoStProof` this function returns. It *does* happen outside the scope of the function which creates it (`createPrivatePoStProof`), and the context of the verification (the `PrivatePoStProof`) never escapes `CreatePoStProof` anyway. The right answer is probably just to delete the comment and the proposed verification, since it's not required for specification and introduces this awkwardness.